### ISSUE TITLE
createMemo should respect the equals value on its initial execution

### ISSUE THREAD

**Issue by snnsnn** [solidjs/solid#2362]
A memo accept an initial value and the `equals` property via the `options` object to exert finer control over its update logic.  However it does not respect the equals property on its initial execution but it should for consistency.

For the following example, ` 'some value'` should never be assigned to the memo as it is not a valid value.

```ts
import { createEffect, createSignal, createMemo } from 'solid-js';

const [preferences, setPreferences] = createSignal({
  lang: 'en', theme: 'some value', fontSize: 'large',
});

const theme = createMemo(() => preferences().theme, 'light', {
  equals: (prev, curr) => {
    if (prev === curr) return true;
    if (prev === 'dark' && curr === 'light') return false;
    if (prev === 'light' && curr === 'dark') return false;
    return true;
  }
});

createEffect(() => console.log(theme()));
```

**Comment by atk** [comment]
But equals by definition compares the last value to the next one. As there is no last value initially, you would have to guard your comparison against it being undefined.

In your case, a filter for invalid values should be done within the derivation, not the equals function.

**Comment by snnsnn** [comment]
I know how it works but what is the point of having initial value if you don't utilize it, if you think about it, it is only logical to use the custom comparator function for the initial execution too. It better aligns with the behavior of a signal.

**Comment by atk** [comment]
createMemo cannot know if you are intending to use the initial value or not.

**Comment by snnsnn** [comment]
> createMemo cannot know if you are intending to use the initial value or not.

Please can you elaborate because I am not sure if I understand what do you mean by this sentence?

If current behaviour is OK, then what is the purpose of the initial value? More importantly what difference does it make if `createMemo` uses the equals value for the initial execution? Memo will be executed any way, equals controls if the returned value should replace the existing one, providing a consistent behavior.

**Comment by atk** [comment]
This was merely an answer to your statement that having an initial value was pointless if you're not using it. But that's merely your preference based on a single case.

In Solid, Ryan usually aims for the most helpful, predictable and performant behavior. Since an initial equality comparison without a previous value will be unhelpful except for very few edge cases, totally unexpected and easily provoking errors and provides the worse performance, he chose not to apply it.

**Comment by ryansolid** [comment]
Yeah it's interesting. I agree that I don't think people expect the equality check to run immediately when no default value is provided. In 2.0 I have been experimenting with running if a default value is provided. But keep in mind we can't tell the difference between there being one and an undefined value so it means undefined would also not run initially. I'd love to have more feedback here in terms of expectations, but I don't see this changing in 1.0 time period.

**Comment by ryansolid** [comment]
I did implement this in 2.0 but it became clear to me as I worked through async I should revert this change. Mostly that the initial value is never seen as the committed value, especially with async. It is a seed value for previous pass in, but it doesn't represent an actual state that is observable from the outside. We should not treat it as such by making it equal.

I'm going to close this as won't fix. But I appreciate this for causing me to look into this deeper.
